### PR TITLE
geo: add st_voronoipolygons and st_voronoilines builtin

### LINKED ISSUE #49065: geo/geomfn: implement ST_VoronoiLines({geometry,float8,geometry})
Implement `ST_VoronoiLines` on arguments {geometry,float8,geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_VoronoiLines.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

The following additional guidance has been issued on implementing this function:

Implemented by GEOS.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-11-15T23:41:27Z. Changes to titles, body and labels may be overwritten.</sub>

### LINKED ISSUE #49066: geo/geomfn: implement ST_VoronoiPolygons({geometry,float8,geometry})
Implement `ST_VoronoiPolygons` on arguments {geometry,float8,geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_VoronoiPolygons.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

The following additional guidance has been issued on implementing this function:

Implemented by GEOS.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-11-15T23:41:27Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by kakshay21**
Resolves https://github.com/cockroachdb/cockroach/issues/49066, Resolves #49065 
Release note (sql change): New builtin method to calculate voronoi polygons
Release note (sql change): New builtin method to calculate voronoi lines

### CODE REVIEW

**Comment by otan** [comment]
hey @kakshay21 -- haven't heard from you in a while. is there anything I can help you with to get this over the line?

**Comment by kakshay21** [comment]
Hey @otan,
Sorry for the delay, I've added [GEOSEqualsExact_r](https://github.com/libgeos/geos/blob/cb2a1dae659825759037e7e81ad16d726e61a3dc/capi/geos_c.h.in#L827) methods to compare two geometry objects with given epsilon. Currently the test is failing for all scenarios when `onlyEdges=0`(see here: [geos_c.h](https://github.com/libgeos/geos/blob/cb2a1dae659825759037e7e81ad16d726e61a3dc/capi/geos_c.h.in#L1809)).
But in the meanwhile, feel free to review the PR.

File: pkg/geo/geos/geos.h [hunk 1]
```diff
@@ -168,6 +168,12 @@ CR_GEOS_Status CR_GEOS_RelateBoundaryNodeRule(CR_GEOS* lib, CR_GEOS_Slice a, CR_
 CR_GEOS_Status CR_GEOS_RelatePattern(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_Slice b,
                                      CR_GEOS_Slice pattern, char* ret);
 
+CR_GEOS_Status CR_GEOS_VoronoiDiagram(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_Slice env,
+                                      double tolerance, int onlyEdges, CR_GEOS_String* ret);
+
+CR_GEOS_Status CR_GEOS_EqualsExact(CR_GEOS* lib, CR_GEOS_Slice lhs, CR_GEOS_Slice rhs,
```
Comment (otan) [inline pkg/geo/geos/geos.h:174]: this should be put in the same category as `Distance`

File: pkg/geo/geos/geos.go [hunk 2]
```diff
@@ -967,3 +967,46 @@ func Node(a geopb.EWKB) (geopb.EWKB, error) {
...
+	var cEWKB C.CR_GEOS_String
+	if err := statusToError(
+		C.CR_GEOS_VoronoiDiagram(g, goToCSlice(a), C.CR_GEOS_Slice{nil, 0}, C.double(tolerance), C.int(onlyEdges), &cEWKB),
+	); err != nil {
+		return nil, err
+	}
+	return cStringToSafeGoBytes(cEWKB), nil
+}
+
+func EqualsExact(lhs, rhs geopb.EWKB, tolerance float64) (bool, error) {
```
Comment (otan) [inline pkg/geo/geos/geos.go:1000]: this should be put in the same category as `Distance`.

File: pkg/geo/geomfn/voronoi.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+package geomfn
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/cockroach/pkg/geo/geos"
+)
+
+// Computes the Voronoi Polygons from the vertices of the supplied geometry.
+func VoronoiPolygons(
+	a geo.Geometry, tolerance float64, onlyEdges int,
```
Comment (otan) [inline pkg/geo/geomfn/voronoi.go:20]: `onlyEdges` should be a bool, converted into a C int at the very last moment (i.e. convert this to an int in geos.go when calling the c library).

File: pkg/geo/geomfn/voronoi.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+) (geo.Geometry, error) {
+	paths, err := geos.VoronoiPolygons(a.EWKB(), tolerance, onlyEdges)
+	gm, err := geo.ParseGeometryFromEWKB(paths)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	return gm, nil
+}
+
+func VoronoiPolygonsWithEnv(
```
Comment (otan) [inline pkg/geo/geomfn/voronoi.go:30]: nit: VoronoiPolygonsWithEnvelope

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 5]
```diff
@@ -5643,6 +5643,69 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
...
+				{"geometry", types.Geometry},
+				{"tolerance", types.Float},
+				{"extend_to", types.Geometry},
+			},
+			ReturnType: tree.FixedReturnType(types.Geometry),
+			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				a := tree.MustBeDGeometry(args[0])
+				tolerance := tree.MustBeDFloat(args[1])
+				b := tree.MustBeDGeometry(args[2])
+				ret, err := geomfn.VoronoiPolygonsWithEnv(a.Geometry, b.Geometry, float64(tolerance), 1)
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5697]: nit: surround with naked literals, i.e. `true /* onlyEdges */`

**Review by otan — COMMENTED** [review]
Can you give more flavour as to which test exactly fails, what and what you expect?

File: pkg/geo/geomfn/voronoi_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+			expected: geo.MustParseGeometry("MULTILINESTRING EMPTY"),
+		},
+		//{
+		//	name: "Computes Voronoi Polygons for a given MultiPoint",
+		//	args: args{
+		//		a: geo.MustParseGeometry("MULTIPOINT ((280 300), (420 330), (380 230), (320 160))"),
+		//		tol: 0,
+		//		poly: 0,
+		//	},
+		//	expected: geo.MustParseGeometry("GEOMETRYCOLLECTION (POLYGON ((110 175.71428571428572, 110 500, 310.35714285714283 500, 353.515625 298.59375, 306.875 231.96428571428572, 110 175.71428571428572)), POLYGON ((590 204, 590 -10, 589.1666666666666 -10, 306.875 231.96428571428572, 353.515625 298.59375, 590 204)), POLYGON ((110 -10, 110 175.71428571428572, 306.875 231.96428571428572, 589.1666666666666 -10, 110 -10)), POLYGON ((310.35714285714283 500, 590 500, 590 204, 353.515625 298.59375, 310.35714285714283 500)))"),
```
Comment (kakshay21) [inline pkg/geo/geomfn/voronoi_test.go:49]: @otan here this test case is failing, that's why I left it commented it for now

Comment (kakshay21) [reply pkg/geo/geomfn/voronoi_test.go:49]: I picked up these tests from `c-deps/geos/tests/unit/capi/GEOSVoronoiDiagramTest.cpp` 😅

Comment (kakshay21) [reply pkg/geo/geomfn/voronoi_test.go:49]: Also, I'll rename poly to `onlyEdges`

Comment (otan) [reply pkg/geo/geomfn/voronoi_test.go:49]: where did you get this result from?
as far as i understand it, we should be using onlyEdges = 0 only when using `ST_VoronoiLines`.

File: pkg/geo/geos/geos.go [hunk 7]
```diff
@@ -967,3 +967,46 @@ func Node(a geopb.EWKB) (geopb.EWKB, error) {
...
+	var cEWKB C.CR_GEOS_String
+	if err := statusToError(
+		C.CR_GEOS_VoronoiDiagram(g, goToCSlice(a), goToCSlice(env), C.double(tolerance), C.int(onlyEdges), &cEWKB),
+		); err != nil {
+		return nil, err
+	}
+	return cStringToSafeGoBytes(cEWKB), nil
+}
+
+func VoronoiPolygons(a geopb.EWKB, tolerance float64, onlyEdges int) (geopb.EWKB, error) {
```
Comment (otan) [inline pkg/geo/geos/geos.go:986]: we should call this VoronoiDiagram -- this library should map to GEOS as directly as possible.

File: pkg/geo/geomfn/voronoi_test.go (hunk 6 above)
Comment (otan) [reply pkg/geo/geomfn/voronoi_test.go:49]: ah nvm, i got that wrong.

**Review by otan — COMMENTED** [review]
Hey @akshayjshah thanks for those improvements - a couple of things before I can merge. You'll need to also ensure CI is green!

**Comment by kakshay21** [comment]
Hey @otan I’m not able to see the CI failures. Seems like it’s still failing, can you help here?

**Comment by otan** [comment]
What happens if you click "Details" on `Github CI (Cockroach`)?

image

**Comment by kakshay21** [comment]
@otan It asks for login credentials

**Comment by kakshay21** [comment]
My bad there was an option for login as guest😓

File: pkg/geo/geomfn/voronoi.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/cockroachdb/cockroach/pkg/geo/geopb"
+	"github.com/cockroachdb/cockroach/pkg/geo/geos"
+)
+
+// VoronoiDiagram Computes the Voronoi Diagram from the vertices of the supplied geometry.
+func VoronoiDiagram(
+	a, env geo.Geometry, tolerance float64, onlyEdges bool,
+) (geo.Geometry, error) {
+	var envEWKB geopb.EWKB
+	if !env.Empty() {
```
Comment (otan) [inline pkg/geo/geomfn/voronoi.go:24]: we should use a pointer for `env` -- an empty geometry may produce different results (e.g. `POLYGON EMPTY`) compared to a 0 value geo.Geometry.
can you also rename all the `a` to `g`?

File: pkg/geo/geos/geos.cc [hunk 9]
```diff
@@ -1359,3 +1369,51 @@ CR_GEOS_Status CR_GEOS_Node(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_String* nodeE
   lib->GEOS_finish_r(handle);
   return toGEOSString(error.data(), error.length());
 }
+
+// Voronoi Diagram
```
Comment (otan) [inline pkg/geo/geos/geos.cc:1373]: you can remove comments from these two  lines -- if you want them, add them in geos.h and explain in more detail about it.

File: pkg/geo/geos/geos.cc [hunk 10]
```diff
@@ -1359,3 +1369,51 @@ CR_GEOS_Status CR_GEOS_Node(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_String* nodeE
...
+// See: https://en.wikipedia.org/wiki/Voronoi_diagram
+CR_GEOS_Status CR_GEOS_VoronoiDiagram(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_Slice env,
```
Comment (otan) [inline pkg/geo/geos/geos.cc:1375]: Can you rename a to g?

File: pkg/geo/geos/geos.h [hunk 11]
```diff
@@ -168,6 +170,9 @@ CR_GEOS_Status CR_GEOS_RelateBoundaryNodeRule(CR_GEOS* lib, CR_GEOS_Slice a, CR_
 CR_GEOS_Status CR_GEOS_RelatePattern(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_Slice b,
                                      CR_GEOS_Slice pattern, char* ret);
 
+CR_GEOS_Status CR_GEOS_VoronoiDiagram(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_Slice env,
```
Comment (otan) [inline pkg/geo/geos/geos.h:173]: nit: rename to `g`

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 12]
```diff
@@ -5643,6 +5643,139 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
...
+	),
+	"st_voronoilines": makeBuiltin(
+		defProps(),
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{"geometry", types.Geometry},
+			},
+			ReturnType: tree.FixedReturnType(types.Geometry),
+			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				a := tree.MustBeDGeometry(args[0])
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5719]: rename a to g

File: pkg/geo/geos/geos.go [hunk 13]
```diff
@@ -983,3 +998,22 @@ func Node(a geopb.EWKB) (geopb.EWKB, error) {
 	}
 	return cStringToSafeGoBytes(cEWKB), nil
 }
+
+// VoronoiDiagram Computes the Voronoi Diagram from the vertices of the supplied EWKBs.
+func VoronoiDiagram(a, env geopb.EWKB, tolerance float64, onlyEdges bool) (geopb.EWKB, error) {
```
Comment (otan) [inline pkg/geo/geos/geos.go:1003]: can you rename `a` to `g`?

File: pkg/geo/geos/geos.go [hunk 14]
```diff
@@ -983,3 +998,22 @@ func Node(a geopb.EWKB) (geopb.EWKB, error) {
...
+	g, err := ensureInitInternal()
+	if err != nil {
+		return nil, err
+	}
+	var cEWKB C.CR_GEOS_String
+	flag := 0
+	if onlyEdges == true {
```
Comment (otan) [inline pkg/geo/geos/geos.go:1010]: nit: `if onlyEdges {`

File: pkg/geo/geos/geos.cc [hunk 15]
```diff
@@ -1359,3 +1369,51 @@ CR_GEOS_Status CR_GEOS_Node(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_String* nodeE
...
+                                      double tolerance, char* ret) {
+  std::string error;
+  auto handle = initHandleWithErrorBuffer(lib, &error);
+  auto lhsGeom = CR_GEOS_GeometryFromSlice(lib, handle, lhs);
+  auto rhsGeom = CR_GEOS_GeometryFromSlice(lib, handle, rhs);
+  *ret = 0;
+  if (lhsGeom != nullptr && rhsGeom != nullptr) {
+    auto r = lib->GEOSEqualsExact_r(handle, lhsGeom, rhsGeom, tolerance);
+    *ret = r;
+    lib->GEOSGeom_destroy_r(handle, lhsGeom);
```
Comment (otan) [inline pkg/geo/geos/geos.cc:1414]: nit:
```
if lhsGeom != nullptr{
       lib->GEOSGeom_destroy_r(handle, lhsGeom);
}
if rhsGeom != nullptr {
       lib->GEOSGeom_destroy_r(handle, rhsGeom);

}
```

**Review by otan — COMMENTED** [review]
Cool -- just a few renames, and if CI is happy I'm happy!

File: pkg/geo/geos/geos.go (hunk 13 above)
Comment (kakshay21) [reply pkg/geo/geos/geos.go:1003]: for the shake of familiarity, I've avoided just this one as there is already `g, err := ensureInitInternal()` in line 1004.

File: pkg/geo/geomfn/voronoi.go (hunk 8 above)
Comment (kakshay21) [reply pkg/geo/geomfn/voronoi.go:24]: right, I guess nil check would be enough?

**Comment by kakshay21** [comment]
hey, @otan did I miss any changes?

**Comment by otan** [comment]
woops, yeah i fixed one small nit but forgot to merge it.

bors r+

**Review by otan — APPROVED** [review]

File: pkg/geo/geos/geos.cc [hunk 16]
```diff
@@ -1359,3 +1369,53 @@ CR_GEOS_Status CR_GEOS_Node(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_String* nodeE
...
+                                      double tolerance, char* ret) {
+  std::string error;
+  auto handle = initHandleWithErrorBuffer(lib, &error);
+  auto lhsGeom = CR_GEOS_GeometryFromSlice(lib, handle, lhs);
+  auto rhsGeom = CR_GEOS_GeometryFromSlice(lib, handle, rhs);
+  *ret = 0;
+  if (lhsGeom != nullptr && rhsGeom != nullptr) {
+    auto r = lib->GEOSEqualsExact_r(handle, lhsGeom, rhsGeom, tolerance);
+    *ret = r;
+    if (lhsGeom != nullptr) {
```
[the remark this answers is not in the document]
Comment (kakshay21) [reply pkg/geo/geos/geos.cc]: Oh 😅
Thank you!

**Comment by SWDevAngel** [comment]
@kakshay21 Hi Kumar. Thank you for contributing to CockroachDB this year. As a token of our appreciation we would like to send you a gift. Please DM me on our [community Slack](https://cockroa.ch/slack) `@lisa` so I can send you a link. (If you don’t want a gift, we also have a charitable contribution choice.) All orders need to be in by December 13, so please send this asap. :)

**Comment by kakshay21** [comment]
Sure @SWDevAngel 
I’ve pinged you with `@akshay` handle.
Thanks!

**Pull request merged**